projects/uriparser/uri_parse_fuzzer.cc

```diff
@@ -0,0 +1,113 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstddef>
+#include <cstring>
+#include <string>
+#include <vector>
+#include <fuzzer/FuzzedDataProvider.h>
+
+using std::string;
+#include "uriparser/include/uriparser/Uri.h"
+#include "uriparser/include/uriparser/UriBase.h"
+
+class UriParserA {
+ public:
+  UriParserA() { memset((void *)&uri_, 0, sizeof(uri_)); }
+  ~UriParserA() { uriFreeUriMembersA(&uri_); }
+
+  UriUriA *get_mutable_uri() { return &uri_; }
+  UriUriA *get_uri() const { return const_cast<UriUriA *>(&uri_); }
+
+ private:
+  UriUriA uri_;
+};
+
+void Escapes(const std::string &uri) {
+  const char *first = uri.c_str();
+  // A new line char takes 6 char to encode.
+  // Use a vector to make a C string.
+  std::vector<char> buf1(uri.size() * 6 + 1);
+  std::vector<char> buf2(uri.size() * 3 + 1);
+
+  char *result = uriEscapeA(first, &buf1[0], URI_TRUE, URI_TRUE);
+  result = uriEscapeA(first, &buf1[0], URI_FALSE, URI_TRUE);
+  result = uriEscapeA(first, &buf2[0], URI_TRUE, URI_FALSE);
+  result = uriEscapeA(first, &buf2[0], URI_FALSE, URI_FALSE);
+}
+
+void FileNames(const std::string &uri) {
+  const size_t size = 8 + 3 * uri.size() + 1;
+  std::vector<char> buf(size);
+}
+
+// Yuck!  The header situation for uriparse is rough.
```

Comment:
How so? Are you are of [this commit](https://github.com/uriparser/uriparser/commit/db539fbed3c3419c5a1ad4bc6e30db43e3800e4a) that is part of uriparser >=0.9.2?

Reply:
Apologies for the snarky comment, thanks for letting us know.

Reply:
I'll also take responsibility for this; we wrote this fuzzer internally at Google when were using a much older version of uriparser.  I should have removed this comment before we released the fuzzer.

Reply:
Thanks for the clarification. I'm happy about bug reports at https://github.com/uriparser/uriparser if you run into cases like that again.